### PR TITLE
usb: device: class: rndis: correct descriptors

### LINKED ISSUE #54574: USB RNDIS Reception Issue(s)
**Describe the bug**

STM32H730VB with embedded USB FS PHY and enabled RNDIS is unstable and doesn't receive all incoming packets.

It feels like `static void rndis_bulk_out(uint8_t ep, enum usb_dc_ep_cb_status_code ep_status)` doesn't receive the complete packet and might corrupt it.

https://github.com/zephyrproject-rtos/zephyr/blob/802600da7b0396ee515ddf68abf69d95ba33ecf5/subsys/usb/device/class/netusb/function_rndis.c#L325-L341

It reads the number of available bytes and sets that as the expected packet length. If there are more than `CONFIG_RNDIS_BULK_EP_MPS` bytes, then we limit the length that we read to `CONFIG_RNDIS_BULK_EP_MPS`. A packet is allocated and filled with `CONFIG_RNDIS_BULK_EP_MPS` of bytes even when `usb_read` returned `len` that is more than `CONFIG_RNDIS_BULK_EP_MPS`. Therefore, when `len` is more than  `CONFIG_RNDIS_BULK_EP_MPS` we discard the data.
The next time the `rndis_bulk_out` is called again we also limit that to `CONFIG_RNDIS_BULK_EP_MPS` and add limited `len` to the packet again and subtract read bytes from the expected packet length.

I observe the following (_on the hardware I have_):

* 1st call to `rndis_bulk_out` with read `len` is 86:
   - The `len` gets limited to 64 (`CONFIG_RNDIS_BULK_EP_MPS=64`). A packet is allocated and filled with 64 bytes.
     The expected length gets set to `86 - 64 = 22`.
* 2nd call to `rndis_bulk_out` with read `len` is 86 (_1st usb_read_):
    - The `len` gets limited to 64 (`CONFIG_RNDIS_BULK_EP_MPS=64`). The packet is filled with additional 64 bytes.
      The expected length gets set to `22 - 64 = -42`.

As soon as the expected packet length is below 0 it is dropped/discarded:

https://github.com/zephyrproject-rtos/zephyr/blob/802600da7b0396ee515ddf68abf69d95ba33ecf5/subsys/usb/device/class/netusb/function_rndis.c#L425-L428

It seems that the current `rndis_bulk_out` implementation works only when received `len` from the 1st `usb_read` is below or equal to `CONFIG_RNDIS_BULK_EP_MPS`.

**Possible Fixes**
Probable fix is to allocate a packet and read the whole payload based on `DataLength` from `REMOTE_NDIS_PACKET_MSG`. Don't just chop and discard the data based on `CONFIG_RNDIS_BULK_EP_MPS`.

Things to consider:

> Each REMOTE_NDIS_PACKET_MSG message contains information about a single network data unit (_such as an Ethernet 802.3 frame_).

and 

> REMOTE_NDIS_PACKET_MSG message on the data channel might contain **multiple** packets. 

> All Windows RNDIS implementations support sending and receiving multiple
REMOTE_NDIS_PACKET_MSG messages in one bus transfer.

Considering:

https://github.com/zephyrproject-rtos/zephyr/blob/0744e42e2299015af339e1530527e5150c1c2621/subsys/usb/device/class/netusb/function_rndis.c#L502-L508

* `MaxPacketsPerTransfer` is 1 
* `MaxTransferSize` is 1558
* `PacketAlignmentFactor` is 0 (2<sup>0</sup>=1)

The `rndis_bulk_out` shall be able to receive transfer up to `1558` bytes (_inclusive_).

**To Reproduce**
1. Configure a project with networking and USB RNDIS for any STM32 USB with FS PHY that sets `CONFIG_RNDIS_BULK_EP_MPS` to `64` (_maybe the same can be observed with HS PHY with `CONFIG_RNDIS_BULK_EP_MPS <= 64`_).
2. Try to receive data on your device.
3. Observe issues when more than `CONFIG_RNDIS_BULK_EP_MPS` data is received within `rndis_bulk_out`.

**Expected behavior**
RNDIS device class shall receive complete packages.

**Impact**
Annoyance and showstopper depending on the RNDIS role and target hardware.

**Logs and console output**
It begins with:
> <wrn> usb_rndis: Limit read len 86 to MPS 64`

Eventually, other log messages might appear (_depending on the application and configured buffer sizes_):
> <err> usb_rndis: Cannot get free buffer
> <wrn> usb_rndis: Transmit response queue is not empty (repeats periodically)
> <wrn> usb_rndis: Notification is already sent

**Environment (please complete the following information):**
 - OS: Windows 10 22H2
 - Toolchain: gcc-arm-none-eabi-10.3-2021.07
 - Zephyr 3.3.0-rc2 https://github.com/zephyrproject-rtos/zephyr/commit/461a13c085de73aabdb5777247a96675472a73ad

**Additional context**
To make RNDIS work on Windows 10 the descriptors require modification so that the device is identified as "RNDIS over Ethernet" and Windows loads the correct driver, see:
* [Windows Driver Supported Classes](https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/supported-usb-classes)
* [USB-IF Defined Class Codes](https://www.usb.org/defined-class-codes)

IAD:
```
.bFunctionClass = USB_BCC_MISCELLANEOUS,
.bFunctionSubClass = 4,
.bFunctionProtocol = 1,
```
IF0:
```
.bInterfaceClass = USB_BCC_MISCELLANEOUS,
.bInterfaceSubClass = 4,
.bInterfaceProtocol = 1,
```

That is discussed in https://github.com/zephyrproject-rtos/zephyr/discussions/51765. I am not sure why https://github.com/zephyrproject-rtos/zephyr/pull/46454 was closed. Created #54611 for descriptors correction.

Furthermore, `CONFIG_USB_DEVICE_OS_DESC` and `CONFIG_USB_COMPOSITE_DEVICE` shall be enabled.

**Specification(s):**
* [RNDIS Specification](https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/WinArchive/[MS-RNDIS].pdf)

### PR DESCRIPTION
**Pull request by GeorgeCGV**
Use descriptors as dictated by Microsoft and USB-IF.

For Remote NDIS using Ethernet medium it shall be:

> Class: Miscellaneous (EFh), SubClass: 04h, Protocol: 01h

That allows the RNDIS device to be correctly detected on Windows.

* [Microsoft](https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/supported-usb-classes)
* [USB-IF Defined Class Codes](https://www.usb.org/defined-class-codes)

There is discussion #51765 related to the issue.
There was an attempt to fix it #46454, but it was closed for some reason.

How it looks without modifications:

as_is_original2

with:

modified_iad_and_ifdesc2

Resolves: #54574

### CODE REVIEW

File: subsys/usb/device/class/netusb/function_rndis.c [hunk 1]
```diff
@@ -62,9 +62,9 @@ USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_rndis_config rndis_cfg = {
 		.bDescriptorType = USB_DESC_INTERFACE_ASSOC,
 		.bFirstInterface = 0,
 		.bInterfaceCount = 0x02,
-		.bFunctionClass = USB_BCC_CDC_CONTROL,
-		.bFunctionSubClass = 6,
-		.bFunctionProtocol = 0,
+		.bFunctionClass = USB_BCC_MISCELLANEOUS,
+		.bFunctionSubClass = 4,
+		.bFunctionProtocol = 1,
```
Comment (jfischer-no) [inline subsys/usb/device/class/netusb/function_rndis.c:67]: The issue with that is, if it is not a CDC class then why should it use CDC descriptors? 
I guess all CDC descriptors below can be omitted, like https://learn.microsoft.com/en-us/azure/rtos/usbx/usbx-device-stack-supplemental-2, still weird.

Comment (GeorgeCGV) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: I am not sure why Microsoft specifies completely different descriptors in that example (_it was noted that in the discussion as well_).

However, such descriptors won't allow Windows to load correct drivers as noted by the screenshots above. At least on Windows 10. Not sure about Windows 11.

Comment (GeorgeCGV) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: Tested with descriptors provided in the example:

ms_descriptors

Windows won't load RNDIS driver automatically. They also mention:

> On windows platforms a .inf file needs to be present on the host that matches the device descriptor.

But using descriptors as specified by USB-IF or by Microsoft in "Supported Device Classes" (_Class: Miscellaneous (EFh), SubClass: 04h, Protocol: 01h_) everything works automatically. Windows loads RNDIS driver.

Comment (tmon-nordic) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: > The issue with that is, if it is not a CDC class then why should it use CDC descriptors?

Why do you think this is CDC descriptor? Class 0xEF, SubClass 0x04, Protocol 0x01 is RNDIS over Ethernet. See https://www.usb.org/defined-class-codes

Comment (GeorgeCGV) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: @tmon-nordic he probably refers to the Microsoft example https://learn.microsoft.com/en-us/azure/rtos/usbx/usbx-device-stack-supplemental-2

Comment (jfischer-no) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: > > The issue with that is, if it is not a CDC class then why should it use CDC descriptors?
> 
> Why do you think this is CDC descriptor? Class 0xEF, SubClass 0x04, Protocol 0x01 is RNDIS over Ethernet. See https://www.usb.org/defined-class-codes

Our is CDC descriptor. Anyway, I think we should just align it with example in https://learn.microsoft.com/en-us/azure/rtos/usbx/usbx-device-stack-supplemental-2 and remove CDC descriptor. @GeorgeCGV can you please test if it also works for you?

```diff
diff --git a/subsys/usb/device/class/netusb/function_rndis.c b/subsys/usb/device/class/netusb/function_rndis.c
index cef31a6a718..8d263563f0c 100644
--- a/subsys/usb/device/class/netusb/function_rndis.c
+++ b/subsys/usb/device/class/netusb/function_rndis.c
@@ -45,10 +45,6 @@ static struct k_thread cmd_thread_data;
 struct usb_rndis_config {
        struct usb_association_descriptor iad;
        struct usb_if_descriptor if0;
-       struct cdc_header_descriptor if0_header;
-       struct cdc_cm_descriptor if0_cm;
-       struct cdc_acm_descriptor if0_acm;
-       struct cdc_union_descriptor if0_union;
        struct usb_ep_descriptor if0_int_ep;
 
```
        struct usb_if_descriptor if1;

Comment (GeorgeCGV) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: @jfischer-no still works on windows with removed `cdc_header_descriptor`, `cdc_cm_descriptor`, `cdc_acm_descriptor`, and `cdc_union_descriptor` while using USB-IF defined class for RNDIS Ethernet.

Comment (jfischer-no) [reply subsys/usb/device/class/netusb/function_rndis.c:67]: > Should I include that in this PR?

Yes, also please add a note to commit message that RNDIS implementation is not a CDC class.

**Review by jfischer-no — APPROVED** [review]

**Review by tmon-nordic — APPROVED** [review]

**Pull request merged**